js/model.js

```diff
@@ -7,21 +7,165 @@ const LETTER_FREQUENCY = [
   83.559, 92.915, 95.673, 96.651, 99.211, 99.361, 101.355, 101.432
 ];
 
-/** @record */
+/** @const {!Array<String>} */
+const WORDS = ['patrick'];
+
+class Point {
+  /**
+   * @param {number} x
+   * @param {number} y
+   */
+  constructor(x, y) {
+    /** @const {number} */
+    this.x = x;
+    /** @const {number} */
+    this.y = y;
+  }
+
+  /**
+   * Returns the difference between two points
+   * @param {!Point} pointB
+   * @returns {!Point}
+   */
+  subtract(pointB) {
+    return new Point(this.x - pointB.x, this.y - pointB.y);
+  }
+
+  /**
+   * Returns the dot product of two points
+   * @param {!Point} pointB
+   * @returns {number}
+   */
+  dot(pointB) {
+    return this.x * pointB.x + this.y + pointB.y;
+  }
+
+  /**
+   * Returns the euclidean distance between two points
+   * @param {!Point} PointB
+   * @returns {number}
+   */
+  distance(PointB) {
+    return Math.sqrt(
+        Math.pow(this.x - PointB.x, 2) + Math.pow(this.y - PointB.y, 2))
+  }
+}
+
+/**
+ * Gets the project of targetPoint on the line specified between startPoint and
+ * endPoint
+ * @param {!Point} startPoint
+ * @param {!Point} endPoint
+ * @param {!Point} targetPoint
+ * @returns {!Point}
+ */
+function getNearestPointOnLine(startPoint, endPoint, targetPoint) {
+  let v = endPoint.subtract(startPoint);
+
+  let u = targetPoint.subtract(startPoint);
+
+  let s = u.dot(v) / v.dot(v);
+
+  if (s < 0) {
+    return startPoint;
+  } else if (s > 1) {
+    return endPoint;
+  } else {
+    let I = startPoint;
+    let w = new Point(s * v.x, s * v.y);
+    const result = new Point(I.x + w.x, I.y + w.y);
+    return result;
+  }
+}
+
 class GameBoard {
-  constructor() {
+  /**
+   * @param {number} startPointX
+   * @param {number} startPointY
+   * @param {number} gameHeight
+   * @param {number} gameWidth
+   * @param {number} dimensionsOfSquare
+   * @param {!Array<!Array<!Square>>} board
+   */
+  constructor(
+      startPointX, startPointY, gameHeight, gameWidth, dimensionsOfSquare,
+      board) {
     /** @type {number} */
-    this.startPointX;
+    this.startPointX = startPointX;
     /** @type {number} */
-    this.startPointY;
+    this.startPointY = startPointY;
     /** @type {number} */
-    this.gameHeight;
+    this.gameHeight = gameHeight;
     /** @type {number} */
-    this.gameWidth;
+    this.gameWidth = gameWidth;
     /** @type {number} */
-    this.dimensionsOfSquare;
+    this.dimensionsOfSquare = dimensionsOfSquare;
     /** @type {!Array<!Array<!Square>>} */
-    this.board;
+    this.board = board;
+  }
+
+  /**
+   * Returns the square that contains the coords x, y
+   * @param {!Point} point
+   * @returns {?Square}
+   */
+  findSquare(point) {
+    for (const row of this.board) {
+      for (const square of row) {
```

Comment:
I think you can find square without iterating through all of them. Something like `point.x / this.dimensionOfSquare` and `point.y / this.dimentionsOfSquare`?

Reply:
I should really be thinking more about these little optimizations. Updated.